compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -4174,37 +4174,76 @@ object Types {
+          }
+
+        def constValue(tp: Type): Option[Any] = tp.fixForEvaluation match {
           case ConstantType(Constant(n)) => Some(n)
           case _ => None
         }
 
-        def boolValue(tp: Type): Option[Boolean] = tp.dealias match {
+        def boolValue(tp: Type): Option[Boolean] = tp.fixForEvaluation match {
           case ConstantType(Constant(n: Boolean)) => Some(n)
           case _ => None
         }
 
-        def intValue(tp: Type): Option[Int] = tp.dealias match {
+        def intValue(tp: Type): Option[Int] = tp.fixForEvaluation match {
           case ConstantType(Constant(n: Int)) => Some(n)
           case _ => None
         }
 
-        def stringValue(tp: Type): Option[String] = tp.dealias match {
-          case ConstantType(Constant(n: String)) => Some(n)
+        def longValue(tp: Type): Option[Long] = tp.fixForEvaluation match {
+          case ConstantType(Constant(n: Long)) => Some(n)
+          case _ => None
+        }
+
+        def floatValue(tp: Type): Option[Float] = tp.fixForEvaluation match {
+          case ConstantType(Constant(n: Float)) => Some(n)
+          case _ => None
+        }
+
+        def doubleValue(tp: Type): Option[Double] = tp.fixForEvaluation match {
+          case ConstantType(Constant(n: Double)) => Some(n)
           case _ => None
         }
 
+        def stringValue(tp: Type): Option[String] = tp.fixForEvaluation match {
+          case ConstantType(Constant(n: String)) => Some(n)
+          case _ => None
+        }
+        def isConst : Option[Type] = args.head.fixForEvaluation match {
+          case ConstantType(_) => Some(ConstantType(Constant(true)))
+          case _ => Some(ConstantType(Constant(false)))
+        }
         def natValue(tp: Type): Option[Int] = intValue(tp).filter(n => n >= 0 && n < Int.MaxValue)
 
         def constantFold1[T](extractor: Type => Option[T], op: T => Any): Option[Type] =
-          extractor(args.head.normalized).map(a => ConstantType(Constant(op(a))))
+          extractor(args.head).map(a => ConstantType(Constant(op(a))))
 
         def constantFold2[T](extractor: Type => Option[T], op: (T, T) => Any): Option[Type] =
+          constantFold2AB(extractor, extractor, op)
+
+        def constantFold2AB[TA, TB](extractorA: Type => Option[TA], extractorB: Type => Option[TB], op: (TA, TB) => Any): Option[Type] =
           for {
-            a <- extractor(args.head.normalized)
-            b <- extractor(args.tail.head.normalized)
+            a <- extractorA(args.head)
+            b <- extractorB(args.last)
           } yield ConstantType(Constant(op(a, b)))
 
+        def constantFold3[TA, TB, TC](
+          extractorA: Type => Option[TA],
+          extractorB: Type => Option[TB],
+          extractorC: Type => Option[TC],
+          op: (TA, TB, TC) => Any
+        ): Option[Type] =
+          for {
+            a <- extractorA(args.head)
+            b <- extractorB(args(1))
+            c <- extractorC(args.last)
```

Comment:
I think I'd be in favor of using `args(0)`, `args(1)` and `args(2)` instead of `.head`, `(1)` and `.last`. I'd want `constantFold3` to throw an exception if it is applied when there are < 3 args. We of course already have protections in place, both with type-checking and the `if nArgs == x` checks, but I think it doesn't hurt to be a tad more defensive.

(same thing in `constantFold2AB`)